### PR TITLE
grpcproxy: return closing error when stream is canceled from conn close

### LINKED ISSUE #6630: TestWatchErrConnClosed fails
https://semaphoreci.com/coreos/etcd/branches/master/builds/1076

[log.txt](https://github.com/coreos/etcd/files/523189/log.txt)

```
=== RUN   TestWatchErrConnClosed
2016-10-11 23:20:02.040819 I | integration: launching node2522201834049211672 (unix://localhost:node2522201834049211672.sock.bridge)
2016-10-11 23:20:02.041431 I | etcdserver: name = node2522201834049211672
2016-10-11 23:20:02.041471 I | etcdserver: data dir = /tmp/etcd506766226
2016-10-11 23:20:02.041512 I | etcdserver: member dir = /tmp/etcd506766226/member
2016-10-11 23:20:02.041550 I | etcdserver: heartbeat = 10ms
2016-10-11 23:20:02.041582 I | etcdserver: election = 100ms
2016-10-11 23:20:02.041611 I | etcdserver: snapshot count = 0
2016-10-11 23:20:02.041649 I | etcdserver: advertise client URLs = unix://127.0.0.1:21524.16121.sock
2016-10-11 23:20:02.041688 I | etcdserver: initial advertise peer URLs = unix://127.0.0.1:21523.16121.sock
2016-10-11 23:20:02.041736 I | etcdserver: initial cluster = node2522201834049211672=unix://127.0.0.1:21523.16121.sock
2016-10-11 23:20:02.060916 I | etcdserver: starting member 2e4b02ca96375a21 in cluster 97a4560d2d323a5f
2016-10-11 23:20:02.060995 I | raft: 2e4b02ca96375a21 became follower at term 0
2016-10-11 23:20:02.061059 I | raft: newRaft 2e4b02ca96375a21 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-10-11 23:20:02.061104 I | raft: 2e4b02ca96375a21 became follower at term 1
2016-10-11 23:20:02.062580 I | etcdserver: set snapshot count to default 10000
2016-10-11 23:20:02.062623 I | etcdserver: starting server... [version: 3.1.0-alpha.1+git, cluster version: to_be_decided]
2016-10-11 23:20:02.081248 I | etcdserver/membership: added member 2e4b02ca96375a21 [unix://127.0.0.1:21523.16121.sock] to cluster 97a4560d2d323a5f
2016-10-11 23:20:02.089753 I | integration: launched node2522201834049211672 (unix://localhost:node2522201834049211672.sock.bridge)
2016-10-11 23:20:02.089834 I | raft: 2e4b02ca96375a21 is starting a new election at term 1
2016-10-11 23:20:02.089904 I | raft: 2e4b02ca96375a21 became candidate at term 2
2016-10-11 23:20:02.089945 I | raft: 2e4b02ca96375a21 received vote from 2e4b02ca96375a21 at term 2
2016-10-11 23:20:02.090001 I | raft: 2e4b02ca96375a21 became leader at term 2
2016-10-11 23:20:02.090046 I | raft: raft.node: 2e4b02ca96375a21 elected leader 2e4b02ca96375a21 at term 2
2016-10-11 23:20:02.096163 W | etcdserver: apply entries took too long [13.066724ms for 1 entries]
2016-10-11 23:20:02.097919 W | etcdserver: avoid queries with large range/delete range!
2016-10-11 23:20:02.098634 I | etcdserver: published {Name:node2522201834049211672 ClientURLs:[unix://127.0.0.1:21524.16121.sock]} to cluster 97a4560d2d323a5f
2016-10-11 23:20:02.098998 I | etcdserver: setting up the initial cluster version to 3.1
2016-10-11 23:20:02.099473 N | etcdserver/membership: set the initial cluster version to 3.1
2016-10-11 23:20:02.115156 I | integration: terminating node2522201834049211672 (unix://localhost:node2522201834049211672.sock.bridge)
2016-10-11 23:20:02.144478 I | integration: terminated node2522201834049211672 (unix://localhost:node2522201834049211672.sock.bridge)
--- FAIL: TestWatchErrConnClosed (0.11s)
    watch_test.go:589: expected grpc: the client connection is closing, got context canceled
```

### PR DESCRIPTION
**Pull request by heyitsanthony**
Fixes #6630

### CODE REVIEW

File: proxy/grpcproxy/watch.go [hunk 1]
```diff
@@ -73,7 +73,12 @@ func (wp *watchProxy) Watch(stream pb.Watch_WatchServer) (err error) {
 	select {
 	case <-wp.ctx.Done():
 		wp.mu.Unlock()
-		return
```
Comment (xiang90) [inline proxy/grpcproxy/watch.go]: yea... we need to kill all naked return :)

File: proxy/grpcproxy/watch.go [hunk 2]
```diff
@@ -73,7 +73,12 @@ func (wp *watchProxy) Watch(stream pb.Watch_WatchServer) (err error) {
...
+		select {
+		case <-wp.leader.disconnectNotify():
+			return grpc.ErrClientConnClosing
+		default:
+			return wp.ctx.Err()
```
Comment (xiang90) [inline proxy/grpcproxy/watch.go]: maybe change line 74 to cerr := <-wp.ctx.Done() and then return cerr here?

**Comment by xiang90** [comment]
LGTM

File: proxy/grpcproxy/watch.go (hunk 2 above)
Comment (heyitsanthony) [inline proxy/grpcproxy/watch.go]: it's a `<-chan struct{}` though...

Comment (xiang90) [inline proxy/grpcproxy/watch.go]: oh. i thought it return error.

**Comment by fanminshi** [comment]
LGTM

**Pull request merged**